pkg/controllers/node/termination/controller.go

```diff
@@ -92,112 +94,193 @@ func (c *Controller) finalize(ctx context.Context, node *corev1.Node) (reconcile
+			// operations on the NodeClaim.
+			return reconcile.Result{RequeueAfter: 1 * time.Second}, nil
+		}
+	} else if !c.hasTerminationGracePeriodElapsed(nodeTerminationTime) {
+		// There are volume attachments blocking instance termination remaining. We should set the status condition to
+		// unknown (if not already) and requeue. This case should never fall through, to continue to instance termination
+		// one of two conditions must be met: all blocking volume attachment objects must be deleted or the nodeclaim's TGP
+		// must have expired.
+		c.recorder.Publish(terminatorevents.NodeAwaitingVolumeDetachmentEvent(node, pendingVolumeAttachments...))
+		stored := nodeClaim.DeepCopy()
+		if modified := nodeClaim.StatusConditions().SetUnknownWithReason(v1.ConditionTypeVolumesDetached, "AwaitingVolumeDetachment", "AwaitingVolumeDetachment"); modified {
+			if err := c.kubeClient.Status().Patch(ctx, nodeClaim, client.MergeFromWithOptions(stored, client.MergeFromWithOptimisticLock{})); err != nil {
+				if errors.IsConflict(err) {
+					return reconcile.Result{Requeue: true}, nil
+				}
+				if errors.IsNotFound(err) {
+					return reconcile.Result{}, c.associatedNodeClaimError(err)
+				}
+				return reconcile.Result{}, err
 			}
-			return reconcile.Result{}, fmt.Errorf("ensuring instance termination, %w", err)
 		}
-		if !isInstanceTerminated {
-			return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+		return reconcile.Result{RequeueAfter: 1 * time.Second}, nil
+	} else {
+		// There are volume attachments blocking instance termination remaining, but the nodeclaim's TGP has expired. In this
+		// case we should set the status condition to false (requeing if it wasn't already) and then fall through to instance
+		// termination.
+		stored := nodeClaim.DeepCopy()
+		if modified := nodeClaim.StatusConditions().SetFalse(v1.ConditionTypeVolumesDetached, "TerminationGracePeriodElapsed", "TerminationGracePeriodElapsed"); modified {
+			if err := c.kubeClient.Status().Patch(ctx, nodeClaim, client.MergeFromWithOptions(stored, client.MergeFromWithOptimisticLock{})); err != nil {
+				if errors.IsConflict(err) {
+					return reconcile.Result{Requeue: true}, nil
+				}
+				if errors.IsNotFound(err) {
+					return reconcile.Result{}, c.associatedNodeClaimError(err)
+				}
+				return reconcile.Result{}, err
+			}
+			// We requeue after a patch operation since we want to ensure we read our own writes before any subsequent
+			// operations on the NodeClaim.
+			return reconcile.Result{RequeueAfter: 1 * time.Second}, nil
 		}
 	}
+
+	isInstanceTerminated, err := termination.EnsureTerminated(ctx, c.kubeClient, nodeClaim, c.cloudProvider)
+	if client.IgnoreNotFound(err) != nil {
+		// 409 - The nodeClaim exists, but its status has already been modified
+		if errors.IsConflict(err) {
+			return reconcile.Result{Requeue: true}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("ensuring instance termination, %w", err)
```

Comment:
We don't seem to handle the NotFound error that we get back from this EnsureTerminated call in the same way as we do elsewhere here

Reply:
I could go either way on this - if we're patching the NodeClaim that means we successfully initiated instance termination. Do we want to orphan the Node at that point because we're unable to patch the NodeClaim? It is an invariant we should maintain, so I could see the argument for consistency.